Add support for Web Extension native messaging via delegate methods.

xeenon opened the pull request:
#### 9ea15e2119dff2b6faf4f5862973b5a88eb37574
<pre>
Add support for Web Extension native messaging via delegate methods.
<a href="https://webkit.org/b/262082">https://webkit.org/b/262082</a>
rdar://problem/116022790

Reviewed by Brady Eidson.

Introduces two new _WKWebExtensionControllerDelegate methods for native messaging.
The first is for single message sending via runtime.sendNativeMessage(), and the other
is for persistent messaging vis runtime.connectNative(), which also introduces the
_WKWebExtensionMessagePort class for handling the port based messaging.

The ultimate plan is to have native messaging work by default internally via NSExtension,
which is tracked by <a href="https://webkit.org/b/262081.">https://webkit.org/b/262081.</a> These delegate methods will still be
useful for testing or for clients that want alternative native messaging.

* Source/WebKit/Modules/OSX_Private.modulemap:
* Source/WebKit/Modules/iOS_Private.modulemap:
* Source/WebKit/Platform/cocoa/CocoaHelpers.h:
(WebKit::isValidJSONObject):
* Source/WebKit/Platform/cocoa/CocoaHelpers.mm:
(WebKit::isValidJSONObject):
(WebKit::encodeJSONData):
* Source/WebKit/Shared/API/APIObject.h:
* Source/WebKit/Shared/Cocoa/APIObject.mm:
(API::Object::newObject):
* Source/WebKit/Shared/Extensions/WebExtensionContentWorldType.h:
(WebKit::toDebugString):
* Source/WebKit/Shared/Extensions/WebExtensionContentWorldType.serialization.in:
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionControllerDelegate.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionMessagePort.h: Added.
(NS_ERROR_ENUM):
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionMessagePort.mm: Added.
(-[_WKWebExtensionMessagePort dealloc]):
(-[_WKWebExtensionMessagePort isEqual:]):
(-[_WKWebExtensionMessagePort applicationIdentifier]):
(-[_WKWebExtensionMessagePort isDisconnected]):
(-[_WKWebExtensionMessagePort sendMessage:completionHandler:]):
(-[_WKWebExtensionMessagePort disconnectWithError:]):
(-[_WKWebExtensionMessagePort _apiObject]):
(-[_WKWebExtensionMessagePort _webExtensionMessagePort]):
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionMessagePortInternal.h: Added.
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionMessagePortPrivate.h: Added.
* Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIPermissionsCocoa.mm:
(WebKit::WebExtensionContext::permissionsRequest):
* Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIPortCocoa.mm:
(WebKit::WebExtensionContext::portPostMessage):
(WebKit::WebExtensionContext::portDisconnect):
(WebKit::WebExtensionContext::removePort):
(WebKit::WebExtensionContext::addNativePort):
(WebKit::WebExtensionContext::removeNativePort):
(WebKit::WebExtensionContext::fireQueuedPortMessageEventsIfNeeded):
(WebKit::WebExtensionContext::sendQueuedNativePortMessagesIfNeeded):
(WebKit::WebExtensionContext::firePortDisconnectEventIfNeeded):
(WebKit::WebExtensionContext::fireQueuedPortMessageEventIfNeeded): Deleted.
* Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIRuntimeCocoa.mm:
(WebKit::WebExtensionContext::runtimeConnect):
(WebKit::WebExtensionContext::runtimeSendNativeMessage):
(WebKit::WebExtensionContext::runtimeConnectNative):
* Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPITabsCocoa.mm:
(WebKit::WebExtensionContext::tabsConnect):
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionMessagePortCocoa.mm: Added.
(WebKit::toAPI):
(WebKit::toWebExtensionMessagePortError):
(WebKit::WebExtensionMessagePort::WebExtensionMessagePort):
(WebKit::WebExtensionMessagePort::operator== const):
(WebKit::WebExtensionMessagePort::extensionContext const):
(WebKit::WebExtensionMessagePort::isDisconnected const):
(WebKit::WebExtensionMessagePort::disconnect):
(WebKit::WebExtensionMessagePort::reportDisconnection):
(WebKit::WebExtensionMessagePort::remove):
(WebKit::WebExtensionMessagePort::sendMessage):
(WebKit::WebExtensionMessagePort::receiveMessage):
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.h:
(WebKit::WebExtensionContext::create):
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.messages.in:
* Source/WebKit/UIProcess/Extensions/WebExtensionMessagePort.h: Added.
(WebKit::WebExtensionMessagePort::create):
(WebKit::WebExtensionMessagePort::~WebExtensionMessagePort):
(WebKit::WebExtensionMessagePort::applicationIdentifier const):
(WebKit::WebExtensionMessagePort::channelIdentifier const):
(WebKit::WebExtensionMessagePort::wrapper const):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIPortCocoa.mm:
(WebKit::WebExtensionAPIPort::sender):
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIRuntimeCocoa.mm:
(WebKit::WebExtensionAPIRuntime::isPropertyAllowed):
(WebKit::WebExtensionAPIRuntime::sendMessage):
(WebKit::WebExtensionAPIRuntime::sendNativeMessage):
(WebKit::WebExtensionAPIRuntime::connectNative):
(WebKit::WebExtensionContextProxy::dispatchRuntimeMessageEvent):
(WebKit::WebExtensionContextProxy::dispatchRuntimeConnectEvent):
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPITabsCocoa.mm:
(WebKit::WebExtensionAPITabs::sendMessage):
* Source/WebKit/WebProcess/Extensions/API/WebExtensionAPIPort.h:
(WebKit::WebExtensionAPIPort::WebExtensionAPIPort):
* Source/WebKit/WebProcess/Extensions/API/WebExtensionAPIRuntime.h:
* Source/WebKit/WebProcess/Extensions/Interfaces/WebExtensionAPIRuntime.idl:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIRuntime.mm:
(TestWebKitAPI::TEST):
* Tools/TestWebKitAPI/cocoa/TestWebExtensionsDelegate.h:
* Tools/TestWebKitAPI/cocoa/TestWebExtensionsDelegate.mm:
(-[TestWebExtensionsDelegate webExtensionController:sendMessage:toApplicationIdentifier:forExtensionContext:replyHandler:]):
(-[TestWebExtensionsDelegate webExtensionController:connectUsingMessagePort:forExtensionContext:completionHandler:]):

Canonical link: <a href="https://commits.webkit.org/268468@main">https://commits.webkit.org/268468@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1fab30b5b8dce51d507ca56720cf447b0a0b36cd

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/19763 "2 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/20186 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/20803 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/21658 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/18469 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/19999 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/23446 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/20330 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/20029 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/19983 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/19983 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/17192 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/22513 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/17156 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/17986 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/24276 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/18231 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/18161 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/22258 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/18751 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/15890 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/17914 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/22265 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/2418 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/18577 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

kiaraarose commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionMessagePortCocoa.mm:
```diff
@@ -0,0 +1,147 @@
+
+void WebExtensionMessagePort::reportDisconnection(Error error)
+{
+    ASSERT(!isDisconnected());
```
why wouldn't the port already be disconnected if we're reporting this it's disconnected?

xeenon replied:
This is the internal method that triggers the disconnect. It should only be called once, so it ASSERTs. There is a public disconnect() that is exposed to JS, and it returns early if already disconnected, since that is the behavior in other browsers and it is safe to call multiple times.

xeenon committed:
Add support for Web Extension native messaging via delegate methods.
https://webkit.org/b/262082
rdar://problem/116022790

Reviewed by Brady Eidson.

Introduces two new _WKWebExtensionControllerDelegate methods for native messaging.
The first is for single message sending via runtime.sendNativeMessage(), and the other
is for persistent messaging vis runtime.connectNative(), which also introduces the
_WKWebExtensionMessagePort class for handling the port based messaging.

The ultimate plan is to have native messaging work by default internally via NSExtension,
which is tracked by https://webkit.org/b/262081. These delegate methods will still be
useful for testing or for clients that want alternative native messaging.

* Source/WebKit/Modules/OSX_Private.modulemap:
* Source/WebKit/Modules/iOS_Private.modulemap:
* Source/WebKit/Platform/cocoa/CocoaHelpers.h:
(WebKit::isValidJSONObject):
* Source/WebKit/Platform/cocoa/CocoaHelpers.mm:
(WebKit::isValidJSONObject):
(WebKit::encodeJSONData):
* Source/WebKit/Shared/API/APIObject.h:
* Source/WebKit/Shared/Cocoa/APIObject.mm:
(API::Object::newObject):
* Source/WebKit/Shared/Extensions/WebExtensionContentWorldType.h:
(WebKit::toDebugString):
* Source/WebKit/Shared/Extensions/WebExtensionContentWorldType.serialization.in:
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionControllerDelegate.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionMessagePort.h: Added.
(NS_ERROR_ENUM):
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionMessagePort.mm: Added.
(-[_WKWebExtensionMessagePort dealloc]):
(-[_WKWebExtensionMessagePort isEqual:]):
(-[_WKWebExtensionMessagePort applicationIdentifier]):
(-[_WKWebExtensionMessagePort isDisconnected]):
(-[_WKWebExtensionMessagePort sendMessage:completionHandler:]):
(-[_WKWebExtensionMessagePort disconnectWithError:]):
(-[_WKWebExtensionMessagePort _apiObject]):
(-[_WKWebExtensionMessagePort _webExtensionMessagePort]):
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionMessagePortInternal.h: Added.
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionMessagePortPrivate.h: Added.
* Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIPermissionsCocoa.mm:
(WebKit::WebExtensionContext::permissionsRequest):
* Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIPortCocoa.mm:
(WebKit::WebExtensionContext::portPostMessage):
(WebKit::WebExtensionContext::portDisconnect):
(WebKit::WebExtensionContext::removePort):
(WebKit::WebExtensionContext::addNativePort):
(WebKit::WebExtensionContext::removeNativePort):
(WebKit::WebExtensionContext::fireQueuedPortMessageEventsIfNeeded):
(WebKit::WebExtensionContext::sendQueuedNativePortMessagesIfNeeded):
(WebKit::WebExtensionContext::firePortDisconnectEventIfNeeded):
(WebKit::WebExtensionContext::fireQueuedPortMessageEventIfNeeded): Deleted.
* Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIRuntimeCocoa.mm:
(WebKit::WebExtensionContext::runtimeConnect):
(WebKit::WebExtensionContext::runtimeSendNativeMessage):
(WebKit::WebExtensionContext::runtimeConnectNative):
* Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPITabsCocoa.mm:
(WebKit::WebExtensionContext::tabsConnect):
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionMessagePortCocoa.mm: Added.
(WebKit::toAPI):
(WebKit::toWebExtensionMessagePortError):
(WebKit::WebExtensionMessagePort::WebExtensionMessagePort):
(WebKit::WebExtensionMessagePort::operator== const):
(WebKit::WebExtensionMessagePort::extensionContext const):
(WebKit::WebExtensionMessagePort::isDisconnected const):
(WebKit::WebExtensionMessagePort::disconnect):
(WebKit::WebExtensionMessagePort::reportDisconnection):
(WebKit::WebExtensionMessagePort::remove):
(WebKit::WebExtensionMessagePort::sendMessage):
(WebKit::WebExtensionMessagePort::receiveMessage):
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.h:
(WebKit::WebExtensionContext::create):
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.messages.in:
* Source/WebKit/UIProcess/Extensions/WebExtensionMessagePort.h: Added.
(WebKit::WebExtensionMessagePort::create):
(WebKit::WebExtensionMessagePort::~WebExtensionMessagePort):
(WebKit::WebExtensionMessagePort::applicationIdentifier const):
(WebKit::WebExtensionMessagePort::channelIdentifier const):
(WebKit::WebExtensionMessagePort::wrapper const):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIPortCocoa.mm:
(WebKit::WebExtensionAPIPort::sender):
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIRuntimeCocoa.mm:
(WebKit::WebExtensionAPIRuntime::isPropertyAllowed):
(WebKit::WebExtensionAPIRuntime::sendMessage):
(WebKit::WebExtensionAPIRuntime::sendNativeMessage):
(WebKit::WebExtensionAPIRuntime::connectNative):
(WebKit::WebExtensionContextProxy::dispatchRuntimeMessageEvent):
(WebKit::WebExtensionContextProxy::dispatchRuntimeConnectEvent):
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPITabsCocoa.mm:
(WebKit::WebExtensionAPITabs::sendMessage):
* Source/WebKit/WebProcess/Extensions/API/WebExtensionAPIPort.h:
(WebKit::WebExtensionAPIPort::WebExtensionAPIPort):
* Source/WebKit/WebProcess/Extensions/API/WebExtensionAPIRuntime.h:
* Source/WebKit/WebProcess/Extensions/Interfaces/WebExtensionAPIRuntime.idl:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIRuntime.mm:
(TestWebKitAPI::TEST):
* Tools/TestWebKitAPI/cocoa/TestWebExtensionsDelegate.h:
* Tools/TestWebKitAPI/cocoa/TestWebExtensionsDelegate.mm:
(-[TestWebExtensionsDelegate webExtensionController:sendMessage:toApplicationIdentifier:forExtensionContext:replyHandler:]):
(-[TestWebExtensionsDelegate webExtensionController:connectUsingMessagePort:forExtensionContext:completionHandler:]):

Canonical link: https://commits.webkit.org/268468@main
